docs/cvl/hooks.md

```diff
@@ -20,77 +22,422 @@ Syntax
+Below is the set of supported opcode hook patterns:
+```cvl
+hook ADDRESS address v
+
+hook BALANCE(address addr) uint v
+
+hook ORIGIN address v
+
+hook CALLER address v
+
+hook CALLVALUE uint v
+
+hook CODESIZE uint v
+
+hook CODECOPY(uint destOffset, uint offset, uint length)
+
+hook GASPRICE uint v
+
+hook EXTCODESIZE(address addr) uint v
+
+hook EXTCODECOPY(address b, uint destOffset, uint offset, uint length)
+
+hook EXTCODEHASH(address a) bytes32 hash
+
+hook BLOCKHASH(uint n) bytes32 hash
+
+hook COINBASE address v
+
+hook TIMESTAMP uint v
+
+hook NUMBER uint v
+
+hook DIFFICULTY uint v
+
+hook GASLIMIT uint v
+
+hook CHAINID uint v
+
+hook SELFBALANCE uint v
+
+hook BASEFEE uint v
+
+hook MSIZE uint v
+
+hook GAS uint v
+
+hook LOG0(uint offset, uint length)
+
+hook LOG1(uint offset, uint length, bytes32 t1)
+
+hook LOG2(uint offset, uint length, bytes32 t1, bytes32 t2)
+
+hook LOG3(uint offset, uint length, bytes32 t1, bytes32 t2, bytes32 t3)
+
+hook LOG4(uint offset, uint length, bytes32 t1, bytes32 t2, bytes32 t3, bytes32 t4)
+
+hook CREATE1(uint value, uint offset, uint length) address v
+
+hook CREATE2(uint value, uint offset, uint length, bytes32 salt) address v 
+
+hook CALL(uint g, address addr, uint value, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc
+
+hook CALLCODE(uint g, address addr, uint value, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc
+
+hook DELEGATECALL(uint g, address addr, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc
+
+hook STATICCALL(uint g, address addr, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc
+
+hook REVERT(uint offset, uint size)
+
+hook SELFDESTRUCT(address a)
+```
+
+% Note: I'm removing the following section and just replacing it by saying that
+% the above list are the only supported codes, since there seem to be many other
+% unsupported codes (e.g. `ADD` and friends)
+% ### Unsupported opcodes
+% 
+% The standard stack-manipulating instructions `DUP*`, `SWAP*`, `PUSH*` and `POP`
+% are not modeled.  `MLOAD` and `MSTORE` are also not modeled.
+
+### Known inter-dependencies and common pitfalls
+
+Hooks are instrumented for every appearance of the matching instruction in the
+bytecode, as generated by a high-level source compiler such as the Solidity
+compiler.  The behavior of the bytecode may sometimes be surprising.  For
+example, every Solidity method call to a non-payable function will contain an
+early call to `CALLVALUE` to check that it is 0. This means that every time a
+non-payable Solidity function is invoked, the `CALLVALUE` hook will be
+triggered.
+
+Hook bodies
+-----------
+
+The body of a hook may contain almost any CVL code, including calls to other
+Solidity functions.  The only exception is that hooks may not contain
+parametric method calls.  Expressions in hook bodies may reference variables
+bound by the hook pattern.
+
+````{todo}
+Questions about the following:
```

Comment:
the core point here is due to summaries being applied before hooks.
so since some summaries like internal summaries are applied before CVL, we get those in the hook bodies.
I suspect that a summary that did not apply early like will also not get applied after hook inlining.

we should also test the 2 cases you mentioned just to make sure our understanding is correct, and to consider which summaries are applied early and which are not (I am mostly suspicious about external summaries).

Reply:
I'm running a bunch of tests now (https://github.com/Certora/EVMVerifier/pull/5288), but based on discussion in #cvl-development I'm pretty sure the doc is accurate now.